File: server/src/main/kotlin/com/duluin/ftth/order/domain/model/Order.kt
package com.duluin.ftth.order.domain.model

import com.duluin.ftth.common.domain.error.ConflictException
import com.duluin.ftth.common.domain.error.ValidationException
import com.duluin.ftth.order.*
import com.duluin.ftth.common.domain.UuidV7
import java.util.UUID

enum class OrderStatus { DRAFT, SUBMITTED, ACCEPTED, SCHEDULED, FULFILLING, FULFILLED, CANCELLED, REJECTED }

/**
 * Penanda yang MENIMPA tampilan status di portal tanpa mengubah status agregat.
 *
 * Dibuat ortogonal terhadap [OrderStatus] (keputusan V184): kalau ia jadi status agregat,
 * setiap mesin transisi — termasuk jalur saga fulfillment — harus tahu cara keluar dari sana,
 * dan setiap penanda baru melipatgandakan pasangan transisi yang harus dijaga.
 *
 * ATURAN BISNIS kapan penanda ini dipasang BELUM diputuskan pemilik produk. Tidak ada satu pun
 * otomasi yang memasangnya; untuk sekarang hanya operator, lewat `POST /api/orders/{id}/attention`.
 */
enum class OrderPortalFlag { WAITING_CUSTOMER, REQUIRES_ATTENTION }

@Suppress("LongParameterList")
class Order private constructor(
    val id: UUID,
    val tenantId: UUID,
    /**
     * Pemesan adalah TEPAT SATU dari [customerId] atau [leadId] — pelanggan yang sudah
     * terdaftar, atau calon pelanggan yang belum. Keduanya terisi berarti satu orang punya
     * dua identitas dan laporan menghitungnya dua kali; keduanya kosong berarti pesanan
     * yatim yang tak bisa ditindaklanjuti. Constraint yang sama ditegakkan di DB (V177).
     */
    val customerId: UUID?,
    val leadId: UUID?,
    /** Nomor yang bisa dibacakan lewat telepon, `ORD-YYMM-NNNN`. Dibuat di luar agregat (lihat OrderNumberGenerator). */
    val orderNumber: String,
    val lines: List<OrderLineCommand>,
    val serviceAddress: ServiceAddress,
    var appointment: Appointment?,
    var status: OrderStatus,
    var cancellationReason: String?,
    var rejectionReason: String?,
    var revision: Long,
    var lastActorId: UUID?,
    var lastOperation: OperationCommand,
    /** Lihat [OrderPortalFlag]. Default null = portal menampilkan status apa adanya. */
    var portalFlag: OrderPortalFlag? = null,
    var portalFlagReason: String? = null,
) {
    companion object {
        fun create(command: CreateOrderCommand, tenantId: UUID, actorId: UUID?, orderNumber: String): Order {
            if (command.lines.isEmpty()) throw ValidationException("Order harus memiliki line")
            command.lines.forEach {
                if (it.quantity <= 0 || it.description.isBlank()) throw ValidationException("Line order tidak valid")
            }
            if ((command.customerId == null) == (command.leadId == null)) {
                throw ValidationException("Order harus punya tepat satu pemesan: pelanggan atau calon pelanggan")
            }
            if (orderNumber.isBlank()) throw ValidationException("Nomor pesanan wajib diisi")
            if (command.serviceAddress.address.isBlank() || command.serviceAddress.city.isBlank() || command.serviceAddress.postalCode.isBlank()) {
                throw ValidationException("Alamat layanan tidak lengkap")
            }
            val latitude = command.serviceAddress.latitude
            val longitude = command.serviceAddress.longitude
            if ((latitude == null) != (longitude == null)) throw ValidationException("Koordinat harus berpasangan")
            if (latitude != null && longitude != null && (latitude !in -90.0..90.0 || longitude !in -180.0..180.0)) {
                throw ValidationException("Koordinat layanan tidak valid")
            }
            command.appointment?.let {
                if (!it.startsAt.isBefore(it.endsAt)) throw ValidationException("Appointment tidak valid")
            }
            return Order(
                UuidV7.generate(), tenantId, command.customerId, command.leadId, orderNumber,
                command.lines.toList(), command.serviceAddress,
                command.appointment, OrderStatus.DRAFT, null, null, 0, actorId, command.operation,
            )
        }

        fun rehydrate(
            id: UUID,
            tenantId: UUID,
            customerId: UUID?,
            leadId: UUID?,
            orderNumber: String,
            lines: List<OrderLineCommand>,
            serviceAddress: ServiceAddress,
            appointment: Appointment?,
            status: OrderStatus,
            cancellationReason: String?,
            rejectionReason: String?,
            revision: Long,
            lastActorId: UUID?,
            lastOperation: OperationCommand,
            portalFlag: OrderPortalFlag? = null,
            portalFlagReason: String? = null,
        ) = Order(id, tenantId, customerId, leadId, orderNumber, lines, serviceAddress, appointment, status,
            cancellationReason, rejectionReason, revision, lastActorId, lastOperation, portalFlag, portalFlagReason)

        private const val MAX_FLAG_REASON = 300

        /** Ujung perjalanan pesanan: tak menunggu siapa pun lagi, jadi tak boleh bertanda. */
        private val TERMINAL = setOf(OrderStatus.FULFILLED, OrderStatus.CANCELLED, OrderStatus.REJECTED)
    }

    fun transition(command: OrderTransitionCommand, actorId: UUID?) {
        if (command.expectedRevision != revision) throw ConflictException("Revision order sudah berubah")
        val target = when (command.transition) {
            OrderTransition.SUBMIT -> requireFrom(OrderStatus.SUBMITTED, OrderStatus.DRAFT)
            OrderTransition.ACCEPT -> requireFrom(OrderStatus.ACCEPTED, OrderStatus.SUBMITTED)
            OrderTransition.SCHEDULE -> requireFrom(OrderStatus.SCHEDULED, OrderStatus.ACCEPTED)
            OrderTransition.START_FULFILLING -> requireFrom(OrderStatus.FULFILLING, OrderStatus.SCHEDULED)
            /*
             * FULFILL diterima juga dari ACCEPTED dan SCHEDULED, bukan hanya FULFILLING.
             *
             * Sejak P5.4/P5.6 hidup pesanan SETELAH diterima dikemudikan oleh work order-nya,
             * bukan oleh operator yang menekan tombol di layar pesanan: approval WO PSB yang
             * memicu transisi ini. Operator TIDAK diwajibkan menekan SCHEDULE lalu
             * START_FULFILLING lebih dulu — dan memang sering tidak, karena penjadwalan
             * sesungguhnya terjadi di work order.
             *
             * Kalau tetap dipaksa lewat FULFILLING, efek `ORDER` di saga akan ditolak dan
             * berakhir sebagai `ORDER_EFFECT_REJECTED` yang menuntut rekonsiliasi manual:
             * pemasangan pelanggan sudah selesai dan sudah disetujui, tapi portalnya
             * selamanya berkata "sedang ditinjau". Kenyataan lapangan yang menang.
             */
            OrderTransition.FULFILL -> requireFrom(
                OrderStatus.FULFILLED, OrderStatus.FULFILLING, OrderStatus.SCHEDULED, OrderStatus.ACCEPTED,
            )
            OrderTransition.CANCEL -> requireFrom(OrderStatus.CANCELLED, OrderStatus.SUBMITTED, OrderStatus.ACCEPTED, OrderStatus.SCHEDULED)
            OrderTransition.REJECT -> requireFrom(OrderStatus.REJECTED, OrderStatus.SUBMITTED)
        }
        if (command.transition == OrderTransition.CANCEL && command.reason.isNullOrBlank()) throw ValidationException("Alasan pembatalan wajib diisi")
        if (command.transition == OrderTransition.REJECT && command.reason.isNullOrBlank()) throw ValidationException("Alasan penolakan wajib diisi")
        if (command.transition == OrderTransition.SCHEDULE && command.appointment == null) throw ValidationException("Appointment wajib diisi")
        status = target
        if (command.transition == OrderTransition.CANCEL) cancellationReason = command.reason
        if (command.transition == OrderTransition.REJECT) rejectionReason = command.reason
        if (command.appointment != null) appointment = command.appointment
        // Pesanan yang sudah sampai ujungnya tak lagi menunggu siapa pun. Penanda yang
        // tertinggal akan membuat portal terus berkata "menunggu Anda" pada pesanan yang
        // pemasangannya sudah selesai — dan pelanggan menelepon menanyakan apa lagi yang kurang.
        if (target in TERMINAL) clearFlag()
        revision += 1
        lastActorId = actorId
        lastOperation = command.operation
    }

    /**
     * Pasang penanda portal. Bukan transisi status: pesanan tetap di tempatnya dan tetap bisa
     * melanjutkan alurnya — yang berubah hanya KALIMAT yang dibaca pelanggan di halaman lacak.
     *
     * [reason] ikut tampil ke pelanggan, jadi ia harus ditulis untuk dibaca pelanggan
     * ("menunggu konfirmasi titik pemasangan"), bukan catatan internal.
     */
    fun flagForPortal(flag: OrderPortalFlag, reason: String?, actorId: UUID?, operation: OperationCommand) {
        if (status in TERMINAL || status == OrderStatus.DRAFT) {
            throw ConflictException("Pesanan berstatus ${status.name} tidak bisa diberi penanda portal")
        }
        val trimmed = reason?.trim()?.ifBlank { null }
        if (trimmed != null && trimmed.length > MAX_FLAG_REASON) {
            throw ValidationException("Alasan penanda maksimal $MAX_FLAG_REASON karakter")
        }
        portalFlag = flag
        portalFlagReason = trimmed
        bump(actorId, operation)
    }

    /** Lepas penanda; pesanan kembali menampilkan status sesungguhnya. */
    fun clearPortalFlag(actorId: UUID?, operation: OperationCommand) {
        if (portalFlag == null) throw ConflictException("Pesanan ini tidak sedang bertanda")
        clearFlag()
        bump(actorId, operation)
    }

    private fun clearFlag() {
        portalFlag = null
        portalFlagReason = null
    }

    private fun bump(actorId: UUID?, operation: OperationCommand) {
        revision += 1
        lastActorId = actorId
        lastOperation = operation
    }

    private fun requireFrom(target: OrderStatus, vararg expected: OrderStatus): OrderStatus {
        if (status !in expected) throw ConflictException("Transisi ${status.name} ke ${target.name} tidak diizinkan")
        return target
    }
}

/**
 * Pemetaan status agregat → kosakata yang aman dibaca PELANGGAN. `null` = pesanan belum layak
 * tampil di portal: DRAFT masih milik operator dan belum tentu pernah benar-benar dipesan.
 *
 * Berdiri sendiri (bukan hanya di dalam [toPortalView]) karena ada DUA pembaca dengan bentuk
 * data berbeda: portal pelanggan yang memegang agregat penuh, dan halaman lacak publik yang
 * hanya memegang satu baris hasil query. Dua salinan aturan ini akan berbeda DIAM-DIAM begitu
 * status baru ditambahkan, dan salah satu dari keduanya akan membocorkan status internal.
 *
 * [flag] MENIMPA status bila ada: pelanggan yang pesanannya menunggu jawabannya sendiri perlu
 * membaca itu, bukan "sedang ditinjau" yang membuatnya menunggu balik.
 */
fun portalStatusOf(status: OrderStatus, flag: OrderPortalFlag?): PortalOrderStatus? {
    val base = when (status) {
        OrderStatus.DRAFT -> null
        OrderStatus.SUBMITTED -> PortalOrderStatus.RECEIVED
        OrderStatus.ACCEPTED -> PortalOrderStatus.REVIEWING
        OrderStatus.SCHEDULED -> PortalOrderStatus.SCHEDULED
        OrderStatus.FULFILLING -> PortalOrderStatus.IN_PROGRESS
        OrderStatus.FULFILLED -> PortalOrderStatus.COMPLETED
        OrderStatus.CANCELLED, OrderStatus.REJECTED -> PortalOrderStatus.CANCELLED
    } ?: return null
    return when (flag) {
        OrderPortalFlag.WAITING_CUSTOMER -> PortalOrderStatus.WAITING_CUSTOMER
        OrderPortalFlag.REQUIRES_ATTENTION -> PortalOrderStatus.REQUIRES_ATTENTION
        null -> base
    }
}

/**
 * Pandangan pesanan yang boleh dilihat PELANGGAN. `null` = pesanan belum layak tampil di
 * portal: DRAFT masih milik operator dan belum tentu pernah benar-benar dipesan.
 *
 * Tinggal di domain (bukan di adapter) karena portal dan konsol harus melihat pemetaan
 * status yang SAMA — dua salinan aturan ini akan berbeda diam-diam begitu status baru
 * ditambahkan, dan pelanggan akan melihat status yang tak pernah dimaksudkan bocor.
 */
fun Order.toPortalView(): PortalOrderView? {
    val portalStatus = portalStatusOf(status, portalFlag) ?: return null
    return PortalOrderView(
        id = id,
        orderNumber = orderNumber,
        status = portalStatus,
        lines = lines.map { OrderLineView(it.catalogItemId, it.description, it.quantity) },
        // Koordinat SENGAJA tidak ikut: portal hanya perlu memastikan alamatnya benar.
        serviceAddress = PortalServiceAddress(serviceAddress.address, serviceAddress.city, serviceAddress.postalCode),
        appointment = appointment,
        revision = revision,
    )
}

File: web/src/api/order.ts
import { api } from './client'
import type { PageResponse } from './types'

/**
 * Pesanan sisi OPERATOR (`/api/orders`). Bentuk datanya diturunkan dari
 * `OrderController`, `OrderLeadController`, dan `OrderImportController` di server.
 *
 * JANGAN disamakan dengan `src/portal/portalClient.ts`: itu permukaan PELANGGAN yang sengaja
 * hanya memuat kosakata aman-dibaca-pelanggan. Berkas ini memuat alamat, alasan penolakan, dan
 * riwayat internal — semuanya hanya untuk konsol berizin `order.*`.
 */

export type OrderStatus =
  | 'DRAFT'
  | 'SUBMITTED'
  | 'ACCEPTED'
  | 'SCHEDULED'
  | 'FULFILLING'
  | 'FULFILLED'
  | 'CANCELLED'
  | 'REJECTED'

export type OrderTransition = 'SUBMIT' | 'ACCEPT' | 'SCHEDULE' | 'START_FULFILLING' | 'FULFILL' | 'CANCEL' | 'REJECT'

/** Penanda yang MENIMPA tampilan status di halaman lacak pelanggan; ortogonal terhadap [OrderStatus]. */
export type OrderPortalFlag = 'WAITING_CUSTOMER' | 'REQUIRES_ATTENTION'

/**
 * SIAPA yang memasang penandanya.
 *
 * Bukan hiasan: penanda SYSTEM dicabut sendiri begitu keadaan yang memicunya berubah, penanda
 * OPERATOR hanya lepas kalau manusia mencabutnya. Tanpa pembedanya operator mencabut penanda
 * sistem, melihatnya muncul lagi beberapa menit kemudian, dan menyimpulkan konsolnya rusak.
 */
export type OrderPortalFlagSource = 'OPERATOR' | 'SYSTEM'

export interface OrderSummaryView {
  readonly id: string
  readonly orderNumber: string
  readonly status: OrderStatus
  readonly customerId: string | null
  readonly leadId: string | null
  /** Null = pemesannya tak bisa diresolusi lagi (mis. pelanggan terhapus); barisnya tetap tampil. */
  readonly requesterName: string | null
  readonly requesterPhone: string | null
  readonly address: string
  readonly city: string
  readonly appointmentStartsAt: string | null
  /** Null = tidak sedang bertanda. Ikut di baris antrean, jadi tak perlu memindai riwayat. */
  readonly portalFlag: OrderPortalFlag | null
  readonly portalFlagReason: string | null
  readonly portalFlagSource: OrderPortalFlagSource | null
  readonly revision: number
  readonly createdAt: string
  readonly updatedAt: string
}

export interface OrderLineView {
  readonly catalogItemId: string
  readonly description: string
  readonly quantity: number
}

export interface OrderServiceAddress {
  readonly address: string
  readonly city: string
  readonly postalCode: string
  readonly latitude: number | null
  readonly longitude: number | null
}

export interface OrderAppointment {
  readonly startsAt: string
  readonly endsAt: string
}

export interface OrderView {
  readonly id: string
  readonly tenantId: string
  readonly customerId: string | null
  /** String polos, bukan [OrderStatus]: server memulangkan `status.name` dari agregat. */
  readonly status: OrderStatus
  readonly lines: OrderLineView[]
  readonly serviceAddress: OrderServiceAddress
  readonly appointment: OrderAppointment | null
  readonly cancellationReason: string | null
  readonly rejectionReason: string | null
  readonly revision: number
  readonly lastActorId: string | null
  readonly leadId: string | null
  readonly orderNumber: string | null
  /**
   * Penanda portal APA ADANYA dari agregat.
   *
   * PERINGATAN: bentuk ini juga dipulangkan sebagai hasil REPLAY operation key yang dibekukan
   * server. Outcome yang dibekukan sebelum ketiga field ini ada akan memulangkan null walau
   * pesanannya bertanda. Jadi jangan pernah menyimpulkan penanda dari respons sebuah tulisan
   * yang mungkin diulang — muat ulang pesanannya lewat [getOrder].
   */
  readonly portalFlag: OrderPortalFlag | null
  readonly portalFlagReason: string | null
  readonly portalFlagSource: OrderPortalFlagSource | null
}

/**
 * Satu baris riwayat pesanan, sumbernya `order_audit`.
 *
 * [toStatus] TIDAK selalu berisi status pesanan: pada `ORDER_FLAGGED` kolom itu dipakai untuk
 * nama PENANDA-nya (lihat `OrderAttentionService`), karena itulah yang berubah di kejadian itu.
 */
export interface OrderTimelineEntryView {
  readonly revision: number
  readonly eventType: string
  readonly fromStatus: string | null
  readonly toStatus: string
  readonly reason: string | null
  readonly actorId: string | null
  readonly occurredAt: string
}

/**
 * Identitas operasi idempoten yang dituntut setiap endpoint tulis pesanan.
 *
 * Nama fieldnya `key`, BUKAN `operationKey` seperti di `fieldservice.ts` — `OperationRequest`
 * di `OrderController` memakai `key`, dan salah nama berarti 400 "key wajib diisi" yang
 * menunjuk ke tempat yang salah.
 */
export interface OrderOperation {
  readonly namespace: string
  readonly key: string
  readonly payloadHash: string
}

/**
 * Bikin identitas operasi baru. Pemanggil WAJIB menyimpannya dan memakai ulang yang SAMA saat
 * mencoba lagi permintaan yang gagal tanpa respons.
 *
 * Alasannya konkret dan mahal: `ACCEPT` bukan sekadar pindah status — ia mempromosikan calon
 * pelanggan dan membuka WO PSB. Operator yang menekan "Terima" lagi karena responsnya hilang di
 * jaringan, dengan kunci baru, akan membuka WO PSB KEDUA untuk pemasangan yang sama, dan teknisi
 * berangkat dua kali ke rumah yang sama.
 */
export function newOrderOperation(namespace: string): OrderOperation {
  return {
    namespace,
    key: crypto.randomUUID(),
    // Hash acak, bukan digest isi payload: server hanya memakainya untuk MEMBANDINGKAN replay
    // kunci yang sama (`payload_hash varchar(128)`), dan pasangan kunci+hash selalu dipakai
    // ulang utuh. Pola yang sama dengan `fieldservice.ts`.
    payloadHash: crypto.randomUUID().replaceAll('-', '').repeat(2),
  }
}

/**
 * Status asal yang sah untuk tiap transisi — cermin `Order.transition` di server.
 *
 * Ada di klien supaya tombol yang PASTI ditolak tidak pernah ditawarkan. Tanpa ini operator
 * menekan "Jadwalkan" pada pesanan yang belum diterima dan hanya mendapat 409 "Transisi
 * SUBMITTED ke SCHEDULED tidak diizinkan" — pesan yang benar, tapi baru muncul setelah ia
 * mengisi janji temunya.
 *
 * FULFILL menerima ACCEPTED dan SCHEDULED SELAIN FULFILLING dengan sengaja: sejak P5.4 hidup
 * pesanan setelah diterima dikemudikan approval WO PSB, dan operator sering tak pernah menekan
 * SCHEDULE/START_FULFILLING sama sekali.
 */
export const ORDER_TRANSITION_SOURCES: Record<OrderTransition, readonly OrderStatus[]> = {
  SUBMIT: ['DRAFT'],
  ACCEPT: ['SUBMITTED'],
  SCHEDULE: ['ACCEPTED'],
  START_FULFILLING: ['SCHEDULED'],
  FULFILL: ['FULFILLING', 'SCHEDULED', 'ACCEPTED'],
  CANCEL: ['SUBMITTED', 'ACCEPTED', 'SCHEDULED'],
  REJECT: ['SUBMITTED'],
}

export function canTransition(status: OrderStatus, transition: OrderTransition): boolean {
  return ORDER_TRANSITION_SOURCES[transition].includes(status)
}

export const ORDER_STATUSES: readonly OrderStatus[] = [
  'DRAFT',
  'SUBMITTED',
  'ACCEPTED',
  'SCHEDULED',
  'FULFILLING',
  'FULFILLED',
  'CANCELLED',
  'REJECTED',
]

export const ORDER_STATUS_LABEL: Record<OrderStatus, string> = {
  DRAFT: 'Draf',
  SUBMITTED: 'Masuk',
  ACCEPTED: 'Diterima',
  SCHEDULED: 'Dijadwalkan',
  FULFILLING: 'Dikerjakan',
  FULFILLED: 'Selesai',
  CANCELLED: 'Dibatalkan',
  REJECTED: 'Ditolak',
}

export const ORDER_TRANSITION_LABEL: Record<OrderTransition, string> = {
  SUBMIT: 'Ajukan',
  ACCEPT: 'Terima',
  SCHEDULE: 'Jadwalkan',
  START_FULFILLING: 'Mulai kerjakan',
  FULFILL: 'Tandai selesai',
  CANCEL: 'Batalkan',
  REJECT: 'Tolak',
}

export const ORDER_PORTAL_FLAG_LABEL: Record<OrderPortalFlag, string> = {
  WAITING_CUSTOMER: 'Menunggu pelanggan',
  REQUIRES_ATTENTION: 'Perlu perhatian',
}

export interface OrderListParams {
  query?: string
  status?: OrderStatus | ''
  createdFrom?: string
  createdTo?: string
  /** `true` = hanya yang bertanda, `false` = hanya yang bersih, tak diisi = semua. */
  flagged?: boolean
  /** Penyaring tanda TERTENTU, dikerjakan server supaya berlaku lintas halaman. */
  portalFlag?: OrderPortalFlag | ''
  page?: number
  size?: number
}

export function listOrders(params: OrderListParams = {}): Promise<PageResponse<OrderSummaryView>> {
  const search = new URLSearchParams()
  if (params.query?.trim()) search.set('query', params.query.trim())
  if (params.status) search.set('status', params.status)
  if (params.createdFrom) search.set('createdFrom', params.createdFrom)
  if (params.createdTo) search.set('createdTo', params.createdTo)
  // `!== undefined`, bukan truthy: `flagged=false` adalah penyaring yang SAH ("hanya yang
  // bersih"). Pemeriksaan truthy diam-diam membuangnya dan memulangkan seluruh antrean.
  if (params.flagged !== undefined) search.set('flagged', String(params.flagged))
  if (params.portalFlag) search.set('portalFlag', params.portalFlag)
  search.set('page', String(params.page ?? 0))
  search.set('size', String(params.size ?? 20))
  return api.get<PageResponse<OrderSummaryView>>(`/api/orders?${search}`)
}

export const getOrder = (id: string) => api.get<OrderView>(`/api/orders/${id}`)

export const getOrderTimeline = (id: string) => api.get<OrderTimelineEntryView[]>(`/api/orders/${id}/timeline`)

export interface OrderTransitionBody {
  readonly expectedRevision: number
  readonly operation: OrderOperation
  readonly reason?: string | null
  readonly appointment?: OrderAppointment | null
  /** Hanya dibaca saat transisinya `ACCEPT`; transisi lain mengabaikannya. */
  readonly promotion?: OrderPromotionBody | null
  readonly assignees?: string[]
}

export interface OrderPromotionBody {
  readonly planId?: string | null
  readonly code?: string | null
  readonly areaId?: string | null
  readonly idCardNumber?: string | null
  readonly monthlyFeeOverride?: string | null
}

export const transitionOrder = (id: string, transition: OrderTransition, body: OrderTransitionBody) =>
  api.post<OrderView>(`/api/orders/${id}/${transition}`, body)

/** `flag: null` MELEPAS penanda. [reason] ikut dibaca PELANGGAN di halaman lacak. */
export const flagOrder = (id: string, body: { flag: OrderPortalFlag | null; reason: string | null; operation: OrderOperation }) =>
  api.post<OrderView>(`/api/orders/${id}/attention`, body)

/** [note] adalah catatan INTERNAL; kalimat untuk pelanggan disusun server, bukan operator. */
export const markOrderUnreachable = (id: string, body: { note: string | null; operation: OrderOperation }) =>
  api.post<OrderView>(`/api/orders/${id}/unreachable`, body)

/**
 * Kapan penanda yang SEDANG terpasang itu dipasang — satu-satunya bagian penanda yang memang
 * hanya ada di riwayat.
 *
 * Dulu SELURUH keadaan penanda direkonstruksi di sini karena server tak memulangkannya sama
 * sekali. Sekarang penandanya ada di `portalFlag`/`portalFlagReason`/`portalFlagSource`, dan yang
 * tersisa hanyalah stempel waktunya. Bedanya penting: fungsi ini TIDAK memutuskan apakah pesanan
 * bertanda — ia hanya menjawab "sejak kapan", dan pemanggilnya wajib sudah tahu jawabannya dari
 * pesanan itu sendiri.
 *
 * Null kalau riwayatnya tak memuat `ORDER_FLAGGED` sama sekali (mis. riwayat gagal dimuat).
 * Pemanggilnya harus tetap menampilkan penandanya tanpa tanggal, bukan menyembunyikannya.
 */
export function portalFlagSetAt(timeline: readonly OrderTimelineEntryView[]): string | null {
  // Riwayat datang urut-naik menurut revisi; pemasangan TERAKHIR yang berlaku.
  for (let index = timeline.length - 1; index >= 0; index -= 1) {
    if (timeline[index].eventType === 'ORDER_FLAGGED') return timeline[index].occurredAt
  }
  return null
}

// ---------------------------------------------------------------------------------------
// Calon pelanggan (lead)
// ---------------------------------------------------------------------------------------

export type LeadStatus = 'NEW' | 'CONTACTED' | 'QUALIFIED' | 'CONVERTED' | 'DROPPED'
export type LeadSource = 'PUBLIC_WEB' | 'CSV_IMPORT' | 'OPERATOR' | 'REFERRAL' | 'OTHER'

export const LEAD_STATUSES: readonly LeadStatus[] = ['NEW', 'CONTACTED', 'QUALIFIED', 'CONVERTED', 'DROPPED']
export const LEAD_SOURCES: readonly LeadSource[] = ['PUBLIC_WEB', 'CSV_IMPORT', 'OPERATOR', 'REFERRAL', 'OTHER']

export const LEAD_STATUS_LABEL: Record<LeadStatus, string> = {
  NEW: 'Baru',
  CONTACTED: 'Sudah dihubungi',
  QUALIFIED: 'Layak',
  CONVERTED: 'Jadi pelanggan',
  DROPPED: 'Batal',
}

export const LEAD_SOURCE_LABEL: Record<LeadSource, string> = {
  PUBLIC_WEB: 'Web publik',
  CSV_IMPORT: 'Impor CSV',
  OPERATOR: 'Operator',
  REFERRAL: 'Referensi',
  OTHER: 'Lainnya',
}

/**
 * Status tujuan yang sah — cermin `OrderLead.changeStatus`. CONVERTED sengaja kosong di semua
 * baris: konversi hanya boleh lewat promosi, karena ia melahirkan pelanggan sungguhan.
 */
export const LEAD_STATUS_TARGETS: Record<LeadStatus, readonly LeadStatus[]> = {
  NEW: ['CONTACTED', 'QUALIFIED', 'DROPPED'],
  CONTACTED: ['QUALIFIED', 'DROPPED'],
  QUALIFIED: ['CONTACTED', 'DROPPED'],
  DROPPED: ['CONTACTED'],
  CONVERTED: [],
}

export interface OrderLeadView {
  readonly id: string
  readonly name: string
  readonly phone: string
  readonly email: string | null
  readonly address: string | null
  readonly latitude: number | null
  readonly longitude: number | null
  readonly interestedPlanId: string | null
  readonly source: LeadSource
  readonly status: LeadStatus
  readonly convertedCustomerId: string | null
  readonly notes: string | null
  readonly createdAt: string
  readonly updatedAt: string
}

export interface PromotedLeadView {
  readonly leadId: string
  readonly customerId: string
  readonly subscriptionId: string | null
  readonly alreadyConverted: boolean
}

export interface OrderLeadListParams {
  query?: string
  status?: LeadStatus | ''
  source?: LeadSource | ''
  page?: number
  size?: number
}

export function listOrderLeads(params: OrderLeadListParams = {}): Promise<PageResponse<OrderLeadView>> {
  const search = new URLSearchParams()
  if (params.query?.trim()) search.set('query', params.query.trim())
  if (params.status) search.set('status', params.status)
  if (params.source) search.set('source', params.source)
  search.set('page', String(params.page ?? 0))
  search.set('size', String(params.size ?? 20))
  return api.get<PageResponse<OrderLeadView>>(`/api/orders/leads?${search}`)
}

export interface CreateOrderLeadBody {
  readonly name: string
  readonly phone: string
  readonly email?: string | null
  readonly address?: string | null
  readonly interestedPlanId?: string | null
  readonly source?: LeadSource
  readonly notes?: string | null
}

export const getOrderLead = (id: string) => api.get<OrderLeadView>(`/api/orders/leads/${id}`)
export const createOrderLead = (body: CreateOrderLeadBody) => api.post<OrderLeadView>('/api/orders/leads', body)
export const changeOrderLeadStatus = (id: string, status: LeadStatus) =>
  api.post<OrderLeadView>(`/api/orders/leads/${id}/status`, { status })
export const promoteOrderLead = (id: string, body: OrderPromotionBody = {}) =>
  api.post<PromotedLeadView>(`/api/orders/leads/${id}/promote`, body)

// ---------------------------------------------------------------------------------------
// Impor CSV massal
// ---------------------------------------------------------------------------------------

export type OrderImportBatchStatus = 'PREVIEWED' | 'COMMITTED'
export type OrderImportRowStatus = 'ACCEPTED' | 'REJECTED' | 'DUPLICATE' | 'CREATED' | 'FAILED'

export interface OrderImportBatchSummaryView {
  readonly id: string
  readonly fileName: string
  readonly status: OrderImportBatchStatus
  readonly delimiter: string
  readonly byteSize: number
  readonly totalRows: number
  readonly acceptedRows: number
  readonly rejectedRows: number
  readonly createdRows: number
  readonly failedRows: number
  readonly importedBy: string | null
  readonly createdAt: string
  readonly committedAt: string | null
}

export interface OrderImportRowView {
  readonly id: string
  readonly lineNumber: number
  readonly status: OrderImportRowStatus
  /** Kalimat untuk MANUSIA — alasan baris ini ditolak. Wajib ditampilkan apa adanya. */
  readonly message: string | null
  readonly name: string | null
  readonly phone: string | null
  readonly email: string | null
  readonly planId: string | null
  readonly address: string | null
  readonly city: string | null
  readonly postalCode: string | null
  readonly notes: string | null
  readonly orderId: string | null
  readonly orderNumber: string | null
}

export interface OrderImportBatchDetailView {
  readonly batch: OrderImportBatchSummaryView
  readonly rows: OrderImportRowView[]
}

export const ORDER_IMPORT_ROW_STATUS_LABEL: Record<OrderImportRowStatus, string> = {
  ACCEPTED: 'Siap dibuat',
  REJECTED: 'Ditolak',
  DUPLICATE: 'Duplikat',
  CREATED: 'Dibuat',
  FAILED: 'Gagal',
}

export const ORDER_IMPORT_BATCH_STATUS_LABEL: Record<OrderImportBatchStatus, string> = {
  PREVIEWED: 'Pratinjau',
  COMMITTED: 'Sudah dijalankan',
}

/** Cermin `OrderImportLimits` di server. Ditampilkan SEBELUM unggah, bukan sesudah ditolak. */
export const ORDER_IMPORT_MAX_BYTES = 2 * 1024 * 1024
export const ORDER_IMPORT_MAX_ROWS = 1000

/** Kolom yang dikenali pengurai; bintang = wajib ada di baris pertama berkas. */
export const ORDER_IMPORT_COLUMNS: readonly { label: string; required: boolean }[] = [
  { label: 'nama', required: true },
  { label: 'hp', required: true },
  { label: 'email', required: false },
  { label: 'paket', required: true },
  { label: 'alamat', required: true },
  { label: 'kota', required: true },
  { label: 'kodepos', required: true },
  { label: 'catatan', required: false },
]

/**
 * Unggah + urai. TIDAK membuat pesanan apa pun — pesanan baru lahir setelah [commitOrderImport].
 *
 * Dikirim lewat `api.postForm` supaya boundary multipart ditulis browser sendiri; memasang
 * `Content-Type` sendiri akan menghasilkan boundary yang tak cocok dan server menolak dengan
 * "Berkas CSV kosong atau tidak terkirim" yang menyesatkan.
 */
export function previewOrderImport(file: File): Promise<OrderImportBatchDetailView> {
  const form = new FormData()
  form.append('file', file)
  return api.postForm<OrderImportBatchDetailView>('/api/orders/import/preview', form)
}

export const getOrderImportBatch = (batchId: string) =>
  api.get<OrderImportBatchDetailView>(`/api/orders/import/${batchId}`)

export const commitOrderImport = (batchId: string) =>
  api.post<OrderImportBatchDetailView>(`/api/orders/import/${batchId}/commit`)

export const listOrderImportHistory = (limit = 20) =>
  api.get<OrderImportBatchSummaryView[]>(`/api/orders/import?limit=${limit}`)

/**
 * Berkas contoh. Diambil lewat `api.blob` (bukan `<a href>`) karena endpoint-nya menuntut
 * header Bearer: tautan biasa tak membawanya dan operator hanya menerima berkas 401.
 */
export const downloadOrderImportTemplate = () => api.blob('/api/orders/import/template')
